deepchem/feat/smiles_tokenizer.py

```diff
@@ -0,0 +1,183 @@
+        return self.ids_to_tokens.get(index, self.unk_token)
+
+    def convert_tokens_to_string(self, tokens):
+        """ Converts a sequence of tokens (string) in a single string. """
+        out_string = " ".join(tokens).replace(" ##", "").strip()
+        return out_string
+
+    def add_special_tokens_ids_single_sequence(self, token_ids):
+        """
+        Adds special tokens to the a sequence for sequence classification tasks.
+        A BERT sequence has the following format: [CLS] X [SEP]
+        """
+        return [self.cls_token_id] + token_ids + [self.sep_token_id]
+
+    def add_special_tokens_single_sequence(self, tokens):
+        """
+        Adds special tokens to the a sequence for sequence classification tasks.
+        A BERT sequence has the following format: [CLS] X [SEP]
+        """
+        return [self.cls_token] + tokens + [self.sep_token]
+
+    def add_special_tokens_sequence_pair(self, token_0, token_1):
+        """
+        Adds special tokens to a sequence pair for sequence classification tasks.
+        A BERT sequence pair has the following format: [CLS] A [SEP] B [SEP]
+        """
+        sep = [self.sep_token]
+        cls = [self.cls_token]
+        return cls + token_0 + sep + token_1 + sep
+
+    def add_special_tokens_ids_sequence_pair(self, token_ids_0, token_ids_1):
+        """
+        Adds special tokens to a sequence pair for sequence classification tasks.
+        A BERT sequence pair has the following format: [CLS] A [SEP] B [SEP]
+        """
+        sep = [self.sep_token_id]
+        cls = [self.cls_token_id]
+        return cls + token_ids_0 + sep + token_ids_1 + sep
+
+    def add_padding_tokens(self, token_ids, length, right=True):
+        """
+        Adds padding tokens to return a sequence of length max_length.
+        By  default padding tokens are added to the right of the sequence.
+        """
+        padding = [self.pad_token_id] * (length - len(token_ids))
+        if right:
+            return token_ids + padding
+        else:
+            return padding + token_ids
+
+    def save_vocabulary(self, vocab_path):
+        """Save the tokenizer vocabulary to a file."""
+        index = 0
+        vocab_file = vocab_path
+        with open(vocab_file, "w", encoding="utf-8") as writer:
+            for token, token_index in sorted(
+                    self.vocab.items(), key=lambda kv: kv[1]
+            ):
+                if index != token_index:
+                    logger.warning(
+                        "Saving vocabulary to {}: vocabulary indices are not consecutive."
+                        " Please check that the vocabulary is not corrupted!".
+                            format(vocab_file)
+                    )
+                    index = token_index
+                writer.write(token + u"\n")
+                index += 1
+        return (vocab_file,)
+
+class BasicSmilesTokenizer(object):
```

Comment:
I think we don't need this class right? It seems most folks would just want `SmilesTokenizer`

Reply:
The SmilesTokenizer uses BasicSmilesTokenizer

Reply:
Ah I see! In that case I suggest renaming this class `_BasicSmilesTokenizer` since this is a private class that shouldn't be part of the public developer facing API

Reply:
It has some functionality, namely that it runs a basic regex SMILES tokenization for those uninterested in using a WordPiece-based tokenization strategy + Huggingface's infrastructure (saving tokenizers, vocab path, etc.). I still think its worth to keep public.